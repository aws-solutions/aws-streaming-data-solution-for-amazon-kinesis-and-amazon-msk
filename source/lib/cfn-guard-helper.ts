/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export class CfnGuardHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: string | string[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push(suppression);
            }
        } else {
            rules.push(suppressions);
        }

        // Ensure that the metadata property exists.
        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        if (resource.cfnOptions.metadata.guard?.SuppressedRules) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.guard.SuppressedRules;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule, rule])
            ).values()
        ];

        resource.cfnOptions.metadata.guard = {
            SuppressedRules: uniqueRules
        }
    }
}
