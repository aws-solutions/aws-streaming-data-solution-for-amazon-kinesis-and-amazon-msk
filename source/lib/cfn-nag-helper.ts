/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */

import * as cdk  from 'aws-cdk-lib';

export interface CfnNagSuppression {
    Id: string;
    Reason: string;
}

export class CfnNagHelper {
    public static addSuppressions(resource: cdk.CfnResource, suppressions: CfnNagSuppression | CfnNagSuppression[]) {
        let rules = [];

        if (suppressions instanceof Array) {
            for (const suppression of suppressions) {
                rules.push({ id: suppression.Id, reason: suppression.Reason });
            }
        } else {
            rules.push({ id: suppressions.Id, reason: suppressions.Reason });
        }

        if (resource.cfnOptions.metadata?.cfn_nag) {
            // If the CfnResource already contains some suppressions, we don't want to erase them.
            const existingRules = resource.cfnOptions.metadata.cfn_nag.rules_to_suppress;
            rules = [...existingRules, ...rules];
        }

        // It's possible that multiple constructs try to add the same suppression.
        // We only keep one occurrence (last) of each.
        // Based on https://stackoverflow.com/a/56768137
        const uniqueRules = [
            ...new Map(
                rules.map(rule => [rule.id, rule])
            ).values()
        ];

        if (!resource.cfnOptions.metadata) {
            resource.cfnOptions.metadata = {};
        }

        resource.cfnOptions.metadata.cfn_nag = {
            rules_to_suppress: uniqueRules
        }
    }
}
