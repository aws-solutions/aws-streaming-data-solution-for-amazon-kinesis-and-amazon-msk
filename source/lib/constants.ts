/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

export default class Constants {
    private static readonly ssmManagedActions = [
        'ssm:DescribeAssociation',
        'ssm:GetDeployablePatchSnapshotForInstance',
        'ssm:GetDocument',
        'ssm:DescribeDocument',
        'ssm:GetManifest',
        'ssm:GetParameter',
        'ssm:GetParameters',
        'ssm:ListAssociations',
        'ssm:ListInstanceAssociations',
        'ssm:PutInventory',
        'ssm:PutComplianceItems',
        'ssm:PutConfigurePackageResult',
        'ssm:UpdateAssociationStatus',
        'ssm:UpdateInstanceAssociationStatus',
        'ssm:UpdateInstanceInformation',
        'ec2messages:AcknowledgeMessage',
        'ec2messages:DeleteMessage',
        'ec2messages:FailMessage',
        'ec2messages:GetEndpoint',
        'ec2messages:GetMessages',
        'ec2messages:SendReply',
        'ssmmessages:CreateControlChannel',
        'ssmmessages:CreateDataChannel',
        'ssmmessages:OpenControlChannel',
        'ssmmessages:OpenDataChannel'
    ]

    public static get SsmManagedActions(): string[] {
        return Constants.ssmManagedActions;
    }
}