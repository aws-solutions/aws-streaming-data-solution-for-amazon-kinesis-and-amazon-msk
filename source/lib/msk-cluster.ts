/*********************************************************************************************************************
 *  Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.                                      *
 *                                                                                                                    *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance    *
 *  with the License. A copy of the License is located at                                                             *
 *                                                                                                                    *
 *      http://www.apache.org/licenses/LICENSE-2.0                                                                    *
 *                                                                                                                    *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES *
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions    *
 *  and limitations under the License.                                                                                *
 *********************************************************************************************************************/

import * as cdk from '@aws-cdk/core';
import * as msk from '@aws-cdk/aws-msk';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as logs from '@aws-cdk/aws-logs';

export interface KafkaClusterProps {
    readonly kafkaVersion: string;
    readonly numberOfBrokerNodes: number;
    readonly brokerInstanceType: string;
    readonly monitoringLevel: string;

    readonly brokerVpcId: string;
    readonly brokerSubnets: string[];
}

export class KafkaCluster extends cdk.Construct {
    private readonly Cluster: msk.CfnCluster;
    private readonly SecurityGroup: ec2.CfnSecurityGroup;

    public get ClusterArn(): string {
        return this.Cluster.ref;
    }

    public get ClusterName(): string {
        return cdk.Fn.join('-', ['kafka-cluster', cdk.Aws.ACCOUNT_ID]);
    }

    public get SecurityGroupId(): string {
        return this.SecurityGroup.ref;
    }

    private MIN_SUBNETS: number = 2;
    private MAX_SUBNETS: number = 3;

    public static get AllowedKafkaVersions(): string[] {
        return ['2.7.0', '2.6.1', '2.6.0', '2.5.1', '2.4.1.1', '2.3.1', '2.2.1'];
    }

    public static get AllowedInstanceTypes(): string[] {
        return ['kafka.m5.large', 'kafka.m5.xlarge', 'kafka.m5.2xlarge', 'kafka.m5.4xlarge', 'kafka.m5.8xlarge', 'kafka.m5.12xlarge', 'kafka.m5.16xlarge', 'kafka.m5.24xlarge', 'kafka.t3.small'];
    }

    public static get AllowedMonitoringLevels(): string[] {
        return ['DEFAULT', 'PER_BROKER', 'PER_TOPIC_PER_BROKER'];
    }

    public static get RequiredRules() {
        return [
            { port: 2181, description: 'ZooKeeper Plaintext' },
            { port: 2182, description: 'ZooKeeper TLS' },
            { port: 9092, description: 'Bootstrap servers Plaintext' },
            { port: 9094, description: 'Bootstrap servers TLS' },
        ];
    }

    constructor(scope: cdk.Construct, id: string, props: KafkaClusterProps) {
        super(scope, id);

        if (!cdk.Token.isUnresolved(props.kafkaVersion) && !KafkaCluster.AllowedKafkaVersions.includes(props.kafkaVersion)) {
            throw new Error(`Unknown Kafka version: ${props.kafkaVersion}`);
        }

        if (!cdk.Token.isUnresolved(props.brokerInstanceType) && !KafkaCluster.AllowedInstanceTypes.includes(props.brokerInstanceType)) {
            throw new Error(`Unknown instance type: ${props.brokerInstanceType}`);
        }

        if (!cdk.Token.isUnresolved(props.monitoringLevel) && !KafkaCluster.AllowedMonitoringLevels.includes(props.monitoringLevel)) {
            throw new Error(`Unknown monitoring level: ${props.monitoringLevel}`);
        }

        if (!cdk.Token.isUnresolved(props.brokerSubnets)) {
            if (props.brokerSubnets.length < this.MIN_SUBNETS || props.brokerSubnets.length > this.MAX_SUBNETS) {
                throw new Error(`brokerSubnets must contain between ${this.MIN_SUBNETS} and ${this.MAX_SUBNETS} items`);
            }
        }

        if (!cdk.Token.isUnresolved(props.numberOfBrokerNodes) && props.numberOfBrokerNodes <= 0) {
            throw new Error('numberOfBrokerNodes must be a positive number');
        }

        if (!cdk.Token.isUnresolved(props.brokerSubnets) && !cdk.Token.isUnresolved(props.numberOfBrokerNodes)) {
            if (props.numberOfBrokerNodes % props.brokerSubnets.length !== 0) {
                throw new Error('numberOfBrokerNodes must be a multiple of brokerSubnets');
            }
        }

        this.SecurityGroup = this.createSecurityGroup(props.brokerVpcId);

        const logGroup = new logs.LogGroup(this, 'LogGroup', { removalPolicy: cdk.RemovalPolicy.RETAIN });
        (logGroup.node.defaultChild as logs.CfnLogGroup).cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'W84',
                    reason: 'Log group data is always encrypted in CloudWatch Logs using an AWS Managed KMS Key'
                }]
            }
        };

        this.Cluster = new msk.CfnCluster(this, 'KafkaCluster', {
            clusterName: this.ClusterName,
            kafkaVersion: props.kafkaVersion,
            numberOfBrokerNodes: props.numberOfBrokerNodes,
            brokerNodeGroupInfo: {
                brokerAzDistribution: 'DEFAULT',
                instanceType: props.brokerInstanceType,
                clientSubnets: props.brokerSubnets,
                securityGroups: [this.SecurityGroupId]
            },
            loggingInfo: {
                brokerLogs: {
                    cloudWatchLogs: {
                        logGroup: logGroup.logGroupName,
                        enabled: true
                    }
                }
            },
            enhancedMonitoring: props.monitoringLevel,
            encryptionInfo: {
                encryptionAtRest: {
                    dataVolumeKmsKeyId: 'alias/aws/kafka'
                },
                encryptionInTransit: {
                    clientBroker: 'TLS',
                    inCluster: true
                }
            },
            openMonitoring: {
                prometheus: {
                    jmxExporter: { enabledInBroker: true },
                    nodeExporter: { enabledInBroker: true }
                }
            }
        });
    }

    private createSecurityGroup(vpcId: string): ec2.CfnSecurityGroup {
        const securityGroup = new ec2.CfnSecurityGroup(this, 'ClusterSG', {
            vpcId: vpcId,
            groupDescription: 'Security group for the MSK cluster',
            tags: [{ key: 'Name', value: 'msk-cluster-sg' }]
        });

        securityGroup.cfnOptions.metadata = {
            cfn_nag: {
                rules_to_suppress: [{
                    id: 'F1000',
                    reason: 'No egress rule defined as default (all traffic allowed outbound) is sufficient for this resource'
                }]
            }
        };

        KafkaCluster.RequiredRules.forEach((rule, index) => {
            new ec2.CfnSecurityGroupIngress(this, `IngressRule${index}`, {
                ipProtocol: 'tcp',
                groupId: securityGroup.ref,
                sourceSecurityGroupId: securityGroup.ref,
                fromPort: rule.port,
                toPort: rule.port,
                description: rule.description
            });
        });

        return securityGroup;
    }
}
