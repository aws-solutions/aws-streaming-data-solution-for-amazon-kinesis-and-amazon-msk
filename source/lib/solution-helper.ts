/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_lambda as lambda } from 'aws-cdk-lib';

import { ExecutionRole } from './lambda-role-cloudwatch';

export interface SolutionHelperProps {
    readonly solutionId: string;
    readonly pattern: string;

    // KDS metrics
    readonly shardCount?: number;
    readonly retentionHours?: number;
    readonly enhancedMonitoring?: string;

    // KDF metrics
    readonly bufferingSize?: number;
    readonly bufferingInterval?: number;
    readonly compressionFormat?: string;

    // MSK metrics
    readonly numberOfBrokerNodes?: number;
    readonly brokerInstanceType?: string;
    readonly monitoringLevel?: string;
    readonly accessControlMethod?: string;
}

export class SolutionHelper extends Construct {
    constructor(scope: Construct, id: string, props: SolutionHelperProps) {
        super(scope, id);

        const metricsMapping = new cdk.CfnMapping(this, 'AnonymousData', {
            mapping: {
                'SendAnonymousData': {
                    'Data': 'Yes'
                }
            }
        });

        const metricsCondition = new cdk.CfnCondition(this, 'AnonymousDatatoAWS', {
            expression: cdk.Fn.conditionEquals(metricsMapping.findInMap('SendAnonymousData', 'Data'), 'Yes')
        });

        const helperRole = new ExecutionRole(this, 'Role');
        const helperFunction = new lambda.Function(this, 'SolutionHelper', {
            runtime: lambda.Runtime.PYTHON_3_12,
            handler: 'lambda_function.handler',
            description: 'This function generates UUID for each deployment and sends anonymous data to the AWS Solutions team',
            role: helperRole.Role,
            code: lambda.Code.fromAsset('lambda/solution-helper'),
            timeout: cdk.Duration.seconds(30)
        });

        const createIdFunction = new cdk.CustomResource(this, 'CreateUniqueID', {
            serviceToken: helperFunction.functionArn,
            properties: {
                'Resource': 'UUID'
            },
            resourceType: 'Custom::CreateUUID'
        });

        const sendDataFunction = new cdk.CustomResource(this, 'SendAnonymousData', {
            serviceToken: helperFunction.functionArn,
            properties: {
                'Resource': 'AnonymousMetric',
                'UUID': createIdFunction.getAttString('UUID'),

                'Region': cdk.Aws.REGION,
                'SolutionId': props.solutionId,
                'Version': '%%VERSION%%',
                'Pattern': props.pattern,

                'ShardCount': props.shardCount,
                'RetentionHours': props.retentionHours,
                'EnhancedMonitoring': props.enhancedMonitoring,

                'BufferingSize': props.bufferingSize,
                'BufferingInterval': props.bufferingInterval,
                'CompressionFormat': props.compressionFormat,

                'NumberOfBrokerNodes': props.numberOfBrokerNodes,
                'BrokerInstanceType': props.brokerInstanceType,
                'MonitoringLevel': props.monitoringLevel,
                'AccessControlMethod': props.accessControlMethod
            },
            resourceType: 'Custom::AnonymousData'
        });

        (helperFunction.node.defaultChild as lambda.CfnFunction).cfnOptions.condition = metricsCondition;
        (createIdFunction.node.defaultChild as lambda.CfnFunction).cfnOptions.condition = metricsCondition;
        (sendDataFunction.node.defaultChild as lambda.CfnFunction).cfnOptions.condition = metricsCondition;
    }
}
