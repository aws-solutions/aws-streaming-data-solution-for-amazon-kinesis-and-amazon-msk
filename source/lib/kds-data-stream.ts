/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_kinesis as kinesis, aws_lambda as lambda, aws_iam as iam } from 'aws-cdk-lib';

import { ExecutionRole } from './lambda-role-cloudwatch';
import { CfnNagHelper } from './cfn-nag-helper';

export interface DataStreamProps {
    readonly shardCount: number;
    readonly retentionPeriod: cdk.Duration;
    readonly enableEnhancedMonitoring: string;
}

export class DataStream extends Construct {
    public readonly Stream: kinesis.Stream;

    constructor(scope: Construct, id: string, props: DataStreamProps) {
        super(scope, id);

        if (!cdk.Token.isUnresolved(props.shardCount) && props.shardCount <= 0) {
            throw new Error('shardCount must be a positive number');
        }

        this.Stream = new kinesis.Stream(this, 'DataStream', {
            encryption: kinesis.StreamEncryption.MANAGED,
            shardCount: props.shardCount,
            retentionPeriod: props.retentionPeriod
        });

        this.createCustomResource(props.enableEnhancedMonitoring);
    }

    private createCustomResource(enableEnhancedMonitoring: string) {
        const customResouceRole = new ExecutionRole(this, 'Role', {
            inlinePolicyName: 'MonitoringPolicy',
            inlinePolicyDocument: new iam.PolicyDocument({
                statements: [new iam.PolicyStatement({
                    resources: ['*'],
                    actions: ['kinesis:EnableEnhancedMonitoring', 'kinesis:DisableEnhancedMonitoring']
                })]
            })
        });

        const cfnRole = customResouceRole.Role.node.defaultChild as iam.CfnRole;
        CfnNagHelper.addSuppressions(cfnRole, {
            Id: 'W11',
            Reason: 'Kinesis enhanced monitoring actions do not support resource level permissions'
        });

        const customResourceFunction = new lambda.Function(this, 'CustomResource', {
            runtime: lambda.Runtime.PYTHON_3_12,
            handler: 'lambda_function.handler',
            role: customResouceRole.Role,
            code: lambda.Code.fromAsset('lambda/kds-enhanced-monitoring'),
            timeout: cdk.Duration.seconds(30)
        });

        new cdk.CustomResource(this, 'EnhancedMonitoring', {
            serviceToken: customResourceFunction.functionArn,
            properties: {
                'EnableEnhancedMonitoring': enableEnhancedMonitoring,
                'StreamName': this.Stream.streamName
            },
            resourceType: 'Custom::EnhancedMonitoring'
        });
    }
}
