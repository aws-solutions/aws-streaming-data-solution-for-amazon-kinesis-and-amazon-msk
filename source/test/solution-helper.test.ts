/** ************************************************************************************************
*   Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.                             *
*   SPDX-License-Identifier: Apache-2.0                                                            *
 ************************************************************************************************ */


import * as cdk  from 'aws-cdk-lib';

import { SolutionHelper } from '../lib/solution-helper';
import { KafkaAccessControl } from '../lib/msk-cluster';

let stack: cdk.Stack;

beforeEach(() => {
    const app = new cdk.App();
    stack = new cdk.Stack(app, 'TestStack');
});

test('creates solution helper without any optional properties', () => {
    new SolutionHelper(stack, 'Helper', {
        solutionId: 'SO9999',
        pattern: 'test-pattern'
    });

});

test('creates solution helper with optional properties', () => {
    new SolutionHelper(stack, 'Helper', {
        solutionId: 'SO9999',
        pattern: 'test-pattern',

        shardCount: 2,
        retentionHours: 24,
        enhancedMonitoring: 'false',

        bufferingSize: 5,
        bufferingInterval: 300,
        compressionFormat: 'GZIP',

        brokerInstanceType: 'kafka.m5.large',
        numberOfBrokerNodes: 2,
        monitoringLevel: 'DEFAULT',
        accessControlMethod: KafkaAccessControl.IAM
    });

});
