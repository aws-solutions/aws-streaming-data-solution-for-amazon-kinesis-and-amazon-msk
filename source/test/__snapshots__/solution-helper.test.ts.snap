// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates solution helper with optional properties 1`] = `
Object {
  "Conditions": Object {
    "HelperAnonymousDatatoAWS6303A68E": Object {
      "Fn::Equals": Array [
        Object {
          "Fn::FindInMap": Array [
            "HelperAnonymousData1341617A",
            "SendAnonymousData",
            "Data",
          ],
        },
        "Yes",
      ],
    },
  },
  "Mappings": Object {
    "HelperAnonymousData1341617A": Object {
      "SendAnonymousData": Object {
        "Data": "Yes",
      },
    },
  },
  "Parameters": Object {
    "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0ArtifactHashB0DC4675": Object {
      "Description": "Artifact hash for asset \\"0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0\\"",
      "Type": "String",
    },
    "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3BucketDFCDF14F": Object {
      "Description": "S3 bucket for asset \\"0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0\\"",
      "Type": "String",
    },
    "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3VersionKeyC50C4EAD": Object {
      "Description": "S3 key for asset version \\"0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HelperCreateUniqueID28DA313C": Object {
      "Condition": "HelperAnonymousDatatoAWS6303A68E",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Resource": "UUID",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "HelperSolutionHelperB4625520",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CreateUUID",
      "UpdateReplacePolicy": "Delete",
    },
    "HelperRole30063D70": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HelperSendAnonymousData42DD297E": Object {
      "Condition": "HelperAnonymousDatatoAWS6303A68E",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccessControlMethod": "IAM access control",
        "BrokerInstanceType": "kafka.m5.large",
        "BufferingInterval": 300,
        "BufferingSize": 5,
        "CompressionFormat": "GZIP",
        "EnhancedMonitoring": "false",
        "MonitoringLevel": "DEFAULT",
        "NumberOfBrokerNodes": 2,
        "Pattern": "test-pattern",
        "Region": Object {
          "Ref": "AWS::Region",
        },
        "Resource": "AnonymousMetric",
        "RetentionHours": 24,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "HelperSolutionHelperB4625520",
            "Arn",
          ],
        },
        "ShardCount": 2,
        "SolutionId": "SO9999",
        "UUID": Object {
          "Fn::GetAtt": Array [
            "HelperCreateUniqueID28DA313C",
            "UUID",
          ],
        },
        "Version": "%%VERSION%%",
      },
      "Type": "Custom::AnonymousData",
      "UpdateReplacePolicy": "Delete",
    },
    "HelperSolutionHelperB4625520": Object {
      "Condition": "HelperAnonymousDatatoAWS6303A68E",
      "DependsOn": Array [
        "HelperRole30063D70",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3BucketDFCDF14F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3VersionKeyC50C4EAD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3VersionKeyC50C4EAD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This function generates UUID for each deployment and sends anonymous data to the AWS Solutions team",
        "Handler": "lambda_function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HelperRole30063D70",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
  },
}
`;

exports[`creates solution helper without any optional properties 1`] = `
Object {
  "Conditions": Object {
    "HelperAnonymousDatatoAWS6303A68E": Object {
      "Fn::Equals": Array [
        Object {
          "Fn::FindInMap": Array [
            "HelperAnonymousData1341617A",
            "SendAnonymousData",
            "Data",
          ],
        },
        "Yes",
      ],
    },
  },
  "Mappings": Object {
    "HelperAnonymousData1341617A": Object {
      "SendAnonymousData": Object {
        "Data": "Yes",
      },
    },
  },
  "Parameters": Object {
    "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0ArtifactHashB0DC4675": Object {
      "Description": "Artifact hash for asset \\"0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0\\"",
      "Type": "String",
    },
    "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3BucketDFCDF14F": Object {
      "Description": "S3 bucket for asset \\"0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0\\"",
      "Type": "String",
    },
    "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3VersionKeyC50C4EAD": Object {
      "Description": "S3 key for asset version \\"0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "HelperCreateUniqueID28DA313C": Object {
      "Condition": "HelperAnonymousDatatoAWS6303A68E",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Resource": "UUID",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "HelperSolutionHelperB4625520",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CreateUUID",
      "UpdateReplacePolicy": "Delete",
    },
    "HelperRole30063D70": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "CloudWatchLogsPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "HelperSendAnonymousData42DD297E": Object {
      "Condition": "HelperAnonymousDatatoAWS6303A68E",
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Pattern": "test-pattern",
        "Region": Object {
          "Ref": "AWS::Region",
        },
        "Resource": "AnonymousMetric",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "HelperSolutionHelperB4625520",
            "Arn",
          ],
        },
        "SolutionId": "SO9999",
        "UUID": Object {
          "Fn::GetAtt": Array [
            "HelperCreateUniqueID28DA313C",
            "UUID",
          ],
        },
        "Version": "%%VERSION%%",
      },
      "Type": "Custom::AnonymousData",
      "UpdateReplacePolicy": "Delete",
    },
    "HelperSolutionHelperB4625520": Object {
      "Condition": "HelperAnonymousDatatoAWS6303A68E",
      "DependsOn": Array [
        "HelperRole30063D70",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3BucketDFCDF14F",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3VersionKeyC50C4EAD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters0713a4b5335df237c2e77cd002ba383ef5a0fe2fbe7ce53ce8645c6cc851d8c0S3VersionKeyC50C4EAD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "This function generates UUID for each deployment and sends anonymous data to the AWS Solutions team",
        "Handler": "lambda_function.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "HelperRole30063D70",
            "Arn",
          ],
        },
        "Runtime": "python3.8",
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
  },
}
`;
